### PR TITLE
Dynamic SSL binding

### PR DESCRIPTION
**Pull request by anisimkov**
TN: UB30-003

### CODE REVIEW

**Review by TurboGit — COMMENTED** [review]
This looks to be going in the right direction. Nice job so far, thanks!

File: config/Makefile [hunk 1]
```diff
@@ -38,20 +38,33 @@ ZPATH = ./
...
 $(SDIR)/src:
 	$(MKDIR) $(SDIR)/src
 
 $(SDIR)/xoscons$(EXEEXT): $(SDIR)/src setup/xoscons.adb
 	-$(GPRBUILD) -p -XPRJ_BUILD=Debug -XLIBRARY_TYPE=static \
 		-XPRJ_TARGET=$(PRJ_TARGET) -XTARGET=$(TARGET) -Psetup xoscons
 
+$(SDIR)/dynamo$(EXEEXT): $(SDIR)/src ssl/dynamo/dynamo.adb
+	-$(GPRBUILD) -p -XPRJ_TARGET=$(PRJ_TARGET) -XTARGET=$(TARGET) \
+		-Pssl/dynamo/dynamo.gpr -g -bargs -E
```
Comment (TurboGit) [inline config/Makefile:59]: Seems like dynamo is missing from this PR now.

File: config/ssl/dynamo/dynamo.adb [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+--                                                                          --
+--  As a special exception, if other files instantiate generics from this   --
+--  unit, or you link this unit with other files to produce an executable,  --
+--  this  unit  does not  by itself cause  the resulting executable to be   --
+--  covered by the GNU General Public License. This exception does not      --
+--  however invalidate any other reasons why the executable file  might be  --
+--  covered by the  GNU Public License.                                     --
+------------------------------------------------------------------------------
+
+with Ada.Command_Line;      use Ada.Command_Line;
```
Comment (TurboGit) [inline config/ssl/dynamo/dynamo.adb:30]: Please move the use clauses into the body.

File: config/ssl/ssl-loader__gnutls.adb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+--                                                                          --
+--  As a special exception, if other files instantiate generics from this   --
+--  unit, or you link this unit with other files to produce an executable,  --
+--  this  unit  does not  by itself cause  the resulting executable to be   --
+--  covered by the GNU General Public License. This exception does not      --
+--  however invalidate any other reasons why the executable file  might be  --
+--  covered by the  GNU Public License.                                     --
+------------------------------------------------------------------------------
+
+with GNATCOLL.Plugins;      use GNATCOLL.Plugins;
```
Comment (TurboGit) [inline config/ssl/ssl-loader__gnutls.adb:30]: use clause in body.

File: config/ssl/ssl-loader__openssl.adb [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+--                                                                          --
+--  As a special exception, if other files instantiate generics from this   --
+--  unit, or you link this unit with other files to produce an executable,  --
+--  this  unit  does not  by itself cause  the resulting executable to be   --
+--  covered by the GNU General Public License. This exception does not      --
+--  however invalidate any other reasons why the executable file  might be  --
+--  covered by the  GNU Public License.                                     --
+------------------------------------------------------------------------------
+
+with GNATCOLL.Plugins;      use GNATCOLL.Plugins;
```
Comment (TurboGit) [inline config/ssl/ssl-loader__openssl.adb:30]: use clause in body.

File: config/ssl/dynamo/dynamo.adb [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+   WLF : constant Wide_Wide_Character :=
+           Wide_Wide_Character'Val (Character'Pos (ASCII.LF));
+
+   Counter : Positive := 1;
+
+   CU : constant Ada_Node := Unit.Root;
+   S  : TIO.File_Type;
+   B  : TIO.File_Type;
+
+   procedure Put (File : TIO.File_Type; Node : Ada_Node'Class);
```
Comment (TurboGit) [inline config/ssl/dynamo/dynamo.adb:70]: We need to comment all routines.
Generally the code is missing comments, can you do a pass and add comments to ensure easier maintenance later.

**Review by TurboGit — CHANGES_REQUESTED** [review]
TIA.

File: config/ssl/dynamo/dynamo.adb (hunk 5 above)
Comment (anisimkov) [reply config/ssl/dynamo/dynamo.adb]: All done. I hope dynamo.adb has enough comments now.

**Review by TurboGit — CHANGES_REQUESTED** [review]
When SSL_DYNAMIC is set but SOCKET=std it fails:

```
$ make SSL_DYNAMIC=true setup
make[1]: Entering directory '/home/obry/dev/repositories/git/aws/config'
Bind
   [gprbind]      xoscons.bexch
   [Ada]          xoscons.ali
Link
   [link]         xoscons.adb
Setup OS specific definitions
Compile
   [Ada]          dynamo.adb
Bind
   [gprbind]      dynamo.bexch
   [Ada]          dynamo.ali
Link
   [link]         dynamo.adb
Setup SSL dynamic binding
Unable to parse file ssl/ssl-thin__std.ads
```

I suppose that in this case we want to issue an error in makefile.check.

File: regtests/0243_sshort/sshort.adb [hunk 6]
```diff
@@ -186,14 +186,18 @@ begin
 
    Peer.Send (Sample (1 .. Last));
 
+   if Peer.Receive /= (1 => 11) then
+      Ada.Text_IO.Put_Line ("Unexpected responce");
```
Comment (TurboGit) [inline regtests/0243_sshort/sshort.adb:190]: `Ada.` can be removed.

Comment (anisimkov) [reply regtests/0243_sshort/sshort.adb]: First issue fixed in separate last commit.
The second one by modify of the same commit.

File: makefile.checks [hunk 7]
```diff
@@ -80,7 +100,16 @@ $(error LaL is set and requires XMLADA to be installed. \
 endif
 endif
 
-ifneq ($(SOCKET), std)
+ifeq ($(SOCKET), std)
+ifeq ($(SSL_DYNAMIC), true)
+# SSL_DYNAMIC can be true only together with SOCKET in openssl | gnutls
+SSL_DYNAMIC = false
```
Comment (TurboGit) [inline makefile.checks:106]: looks like some debug code here, no?

Comment (anisimkov) [reply makefile.checks:106]: No. It mean that if SSL_DYNAMIC defined in makefile.setup as true and not defined in command line then we turn it into false. Next line is comparison SSL_DYNAMIC with true. The SSL_DYNAMIC can be true after SSL_DYNAMIC = false only if the SSL_DYNAMIC=true defined in command line.

**Review by TurboGit — APPROVED** [review]
Ok, looks good. I have approved but we need a bit of documentation for this in the User's Manual. Can you propose something and I can edit and integrate?

**Pull request merged**